### PR TITLE
Add abi.encodeError

### LINKED ISSUE #14287: Add `abi.encodeError` similarly to `abi.encodeCall`
## Abstract

Using `abi.encodeCall` for encoding errors fails with error
```
Error (3509): Expected regular external function type, or external view on public function. Cannot use errors for abi.encodeCall.
```

It would be nice to have a `abi.encodeError` that works similarly to `abi.encodeCall`, but for custom errors.

## Motivation

Both functions and error include a `.selector` getter that returns the `bytes4` "signature". In the case of a function selector, this can be used, along with `abi.encodeWithSelector` to generate the bytes corresponding to a function call. Similarly, `abi.encodeWithSelector` can be used to encode a custom error.

`abi.encodeCall` is a safer alternative that verifies the type and number or arguments at compile type. While `abi.encodeCall` works well with functions, it does not currently support errors.

Being able to generate the bytes corresponding to custom errors can be particularly usefull for testing with Foundry.

## Specification

Given a custom error
```solidity
error SomeErrorName(address arg0, uint256 arg1, bytes4 arg2, bytes32 arg3);
```

Enable the syntax
```solidity
abi.encodeError(SomeErrorName, (arg0, arg1, arg2, arg3));
```
that returns the same as
```solidity
abi.encodeWithSelector(SomeErrorName.selector, arg0, arg1, arg2, arg3);
```
with the same type checking as `abi.encodeCall`

## Backwards Compatibility

None. This is introducing a new syntax that doesn't conflict with existing ones.

### PR DESCRIPTION
**Pull request by Amxx**
Fixes #14287

TODO:
- [ ] code deducpliation between `typeCheckABIEncodeCallFunction` and `typeCheckABIEncodeErrorFunction` ?
  - I'm not sure if there is a clean way of doing that. Considering there is already a lot of code duplication in the codebase, maybe that is ok.
- [x] tests
- [ ] review

### CODE REVIEW

File: test/libsolidity/syntaxTests/abiEncoder/abi_encodeError_free_func.sol [hunk 1]
```diff
@@ -0,0 +1,7 @@
+function g(uint) {}
+
+function f() {
+    abi.encodeError(g, (1));
+}
+// ----
+// TypeError 3510: (56-57): Expected an error type. Cannot use functions for abi.encodeError.
```
Comment (nikola-matic) [inline test/libsolidity/syntaxTests/abiEncoder/abi_encodeError_free_func.sol:7]: Needs another case for library functions.

**Review by nikola-matic — COMMENTED** [review]
Needs some semantic tests as well. A changelog entry as well.

**Comment by r0qs** [comment]
This CI error: https://app.circleci.com/pipelines/github/ethereum/solidity/33577/workflows/8c3156a6-35f8-4aeb-b6e3-18f80f4cb6cf/jobs/1512784 is because the test file name is wrong, see: `test/libsolidity/syntaxTests/specialFunctions/encodeCall_fail_args_internal_function_pointer_for_uint copy.sol`. There is a `_copy` in the filename of this test https://github.com/ethereum/solidity/pull/14974/files#diff-41100eaa90ef551806d2098a1c4f80fe0f86d9587edbfb6f1161c6d543f1f92dR1.

**Comment by Amxx** [comment]
> Needs some semantic tests as well. A changelog entry as well.

Yes, i'm still working on the testing part.

Edit: should be good now

**Comment by Amxx** [comment]
Apparently, in the semantic tests, the returned `bytes memory` are processed in chunks of 32 bytes. So when the buffer has `length ≡ 4 mod 0x20`, we need to pad the expected output with zeros (28 bytes worth of zero exactly)

Is that supposed to happen?

**Comment by cameel** [comment]
Could be just another instance of #13989. The issue is about fixed bytes, but maybe it affects `bytes` too.
I think that soltest is just not padding byte arrays correctly.

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -2445,6 +2451,131 @@ void TypeChecker::typeCheckABIEncodeCallFunction(FunctionCall const& _functionCa
...
+	}
+	solAssert(!externalFunctionType->takesArbitraryParameters(), "Function must have fixed parameters.");
+	// Tuples with only one component become that component
+	std::vector<ASTPointer<Expression const>> callArguments;
+
+	auto const* tupleType = dynamic_cast<TupleType const*>(type(*arguments[1]));
+	if (tupleType)
+	{
+		if (TupleExpression const* argumentTuple = dynamic_cast<TupleExpression const*>(arguments[1].get()))
+			callArguments = decltype(callArguments){argumentTuple->components().begin(), argumentTuple->components().end()};
```
Comment (matheusaaguiar) [inline libsolidity/analysis/TypeChecker.cpp]:
```suggestion
			callArguments = {argumentTuple->components().begin(), argumentTuple->components().end()};
```
I think this is better. `std::initializer_list` gets converted automatically.

File: libsolidity/analysis/TypeChecker.cpp [hunk 3]
```diff
@@ -2227,12 +2228,17 @@ void TypeChecker::typeCheckABIEncodeFunctions(
...
 	typeCheckFunctionGeneralChecks(_functionCall, _functionType);
 
-	// No further generic checks needed as we do a precise check for ABIEncodeCall
+	// No further generic checks needed as we do a precise check for ABIEncodeCall and ABIEncodeError
 	if (_functionType->kind() == FunctionType::Kind::ABIEncodeCall)
 	{
 		typeCheckABIEncodeCallFunction(_functionCall);
 		return;
 	}
+	else if (_functionType->kind() == FunctionType::Kind::ABIEncodeError)
```
Comment (matheusaaguiar) [inline libsolidity/analysis/TypeChecker.cpp]:
```suggestion
	if (_functionType->kind() == FunctionType::Kind::ABIEncodeError)
```

File: libsolidity/analysis/TypeChecker.cpp [hunk 4]
```diff
@@ -2445,6 +2451,131 @@ void TypeChecker::typeCheckABIEncodeCallFunction(FunctionCall const& _functionCa
...
+				msg += " Provided creation function.";
+				break;
+			case FunctionType::Kind::Event:
+				msg += " Cannot use events for abi.encodeError.";
+				break;
+			default:
+				msg += " Cannot use special function.";
+		}
+
+		m_errorReporter.typeError(3510_error, arguments[0]->location(), {}, msg);
```
Comment (matheusaaguiar) [inline libsolidity/analysis/TypeChecker.cpp]:
```suggestion
		m_errorReporter.typeError(3510_error, arguments[0]->location(), msg);
```
We also have one overload which accepts just one location.

**Review by matheusaaguiar — COMMENTED** [review]
This needs documentation entries in https://docs.soliditylang.org/en/v0.8.23/cheatsheet.html and https://docs.soliditylang.org/en/v0.8.23/units-and-global-variables.html#abi-encoding-and-decoding-functions (at least, there may be other places, but those are the main ones I guess).
I think the implementation is good, just left some minor suggestions. However, I think we could avoid the duplicated code with some `if`s and ternary conditions and have it all in one `typeCheckABIEncodeCallFunctionOrError` (or something like that), but I don't consider that should impede this PR from being merged.

**Comment by Amxx** [comment]
@matheusaaguiar 
- I added some references in the files you mentionned
- I fixed the merge conflicts and fixed what needed to be fixed
- I also merged previous functions into `typeCheckABIEncodeCallFunctionOrError` (added the type object and some switches)

File: docs/cheatsheet.rst [hunk 5]
```diff
@@ -26,6 +26,8 @@ ABI Encoding and Decoding Functions
   tuple. Performs a full type-check, ensuring the types match the function signature. Result equals ``abi.encodeWithSelector(functionPointer.selector, (...))``
 - ``abi.encodeWithSignature(string memory signature, ...) returns (bytes memory)``: Equivalent
   to ``abi.encodeWithSelector(bytes4(keccak256(bytes(signature))), ...)``
+- ``abi.encodeError(error errorPointer, (...)) returns (bytes memory)``: ABI-encodes the revert data for ``errorPointer`` with the arguments found in the
+  tuple. Performs a full type-check, ensuring the types match the error definition. Results equals ``abi.encodeWithSelector(errorPointer.selector, (...))``
```
Comment (matheusaaguiar) [inline docs/cheatsheet.rst]:
```suggestion
  tuple. Performs a full type-check, ensuring the types match the error definition. Results equals ``abi.encodeWithSelector(errorPointer.selector, ...)``
```
This a minor imprecision that was fixed recently in the previous entry for `abi.encodeCall` :)
~~I will already commit this suggestion via github interface;~~
Actually, there were another instances which I already fixed via a commit.

File: libsolidity/analysis/TypeChecker.cpp [hunk 6]
```diff
@@ -2316,17 +2335,32 @@ void TypeChecker::typeCheckABIEncodeCallFunction(FunctionCall const& _functionCa
...
+					"Expected first argument to be a function pointer, not \"" +
+					type(*arguments.front())->humanReadableName() +
+					"\"."
+				);
+				return;
+			case FunctionType::Kind::ABIEncodeError:
+				m_errorReporter.typeError(
+					5512_error,
+					arguments.front()->location(),
+					"Expected first argument to be a custom error, not \"" +
```
Comment (matheusaaguiar) [inline libsolidity/analysis/TypeChecker.cpp:2352]: I was trying to think of some test cases and when I try something like:
```solidity
error E(uint a, uint b);
contract C {
    function g() public view returns (bytes memory) {
        bytes memory b1 = abi.encodeError(E(1, 2), (1, 2));
        return b1;
    }
}
```
The following error message is generated:
`Expected first argument to be a custom error, not "error".`
Which is a bit weird. Maybe we should be more precise and change to `[...] custom error name [...]` ?
I think we should have a test like this anyway.

**Review by matheusaaguiar — APPROVED** [review]
I just have one nitpick left, but overall looks good to me.
Also, I forgot to mention that we use `rebase`  rather than `merge` and this needs the former.

**Comment by ekpyron** [comment]
Pending the discussion in https://github.com/ethereum/solidity/issues/14287#issuecomment-2096501517, we decided to close this PR for now. But feel free to jump in there (or join one of our calls) and make the case for this approach and we can potentially revisit and revive this PR!
But as a general policy, we should discuss and confirm the approach before implementations, to be sure we're on the same page about it.

**Comment by Amxx** [comment]
@ekpyron [This change to ERC-7751](https://ethereum-magicians.org/t/erc-7751-wrapping-of-bubbled-up-reverts/20740/3?u=amxx), if it was to happen (note, I'm a co-author of the ERC, so not unlikelly), would create a significant usecase of `abi.encodeError`.

Depending on the feedback this gets, I'd like to reconsider this feature.

**Pull request closed without merging**